src/inmanta/deploy/scheduler.py

```diff
@@ -298,25 +301,95 @@ async def _initialize(self) -> None:
-            up_to_date_resources: Set[ResourceIdStr]
-            _, up_to_date_resources = await ConfigurationModel.get_increment(self.environment, version, connection=con)
+            # Check at which model version the scheduler was before it went down
+            scheduler: Optional[data.Scheduler] = await data.Scheduler.get_one(environment=self.environment, connection=con)
+            assert scheduler is not None
+            last_processed_model_version = scheduler.last_processed_model_version
+            if last_processed_model_version is not None:
+                # Restore scheduler state like it was before the scheduler went down
+                self._state = await ModelState.create_from_db(self.environment, last_processed_model_version, connection=con)
+                self._work.link_to_new_requires_provides_view(
+                    requires_view=self._state.requires.requires_view(),
+                    provides_view=self._state.requires.provides_view(),
+                )
+            else:
+                # This case can occur in two different situations:
+                #   * A model version has been released, but the scheduler didn't process any version yet.
+                #     In this case there is no scheduler state to restore.
+                #   * We migrated the Inmanta server from an old version, that didn't have the resource state
+                #     tracking in the database, to a version that does. To cover this case, we rely on the
+                #     increment calculation to determine which resources have to be considered dirty and which
+                #     not. This migration code path can be removed in a later major version.
+                await self._recover_scheduler_state_using_increments_calculation(connection=con)
+
+            if last_processed_model_version is not None and last_processed_model_version < latest_release_model.version:
+                # We restored the scheduler state from the database, but a newer, released version is available.
+                # Apply the changes from each subsequent model version to the scheduler.
+                versions = await data.ConfigurationModel.get_released_versions_in_interval(
+                    environment=self.environment,
+                    lower_bound=last_processed_model_version + 1,
+                    upper_bound=latest_release_model.version,
+                    connection=con,
+                )
+                for version in reversed(versions):
+                    # Apply version from old to new
+                    await self._read_version(
+                        version=version,
+                        start_deployment=False,
+                        intent_lock_acquired=True,
+                        scheduler_lock_acquired=True,
+                        connection=con,
+                    )
+            # Start deploying everything in dirty set.
+            self._work.deploy_with_context(
```

Comment:
Additionally, `_new_version` assumes that timers are running for everything that is not scheduled when it's called, which is not the case if we restore scheduler state but not timer state.

Then again, #8453 will affect this as well. Proposal: we take the path of least resistance to get this merged, and I rework `read_version` in a follow-up PR to also do the iterative approach, which may unify the two flows again. For now, that means to start timers here (`timer_manager.update_timers()`) for all resources not in the dirty set.

Reply:
Path of least resistance is probably to revert the iteration part of this PR. Drop the associated extra parameters from `_read_version` etc, and simply call into the method as any `read_version` call would. Then we can dedicate a separate PR to the iterative approach, in a unified manner rather than to special-case it for the initializiation.